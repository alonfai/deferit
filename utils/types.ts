export enum Status {
  Processing = 'This bill is currently in processing, it can take approx. 1-2 hours depending on the time of day',
  Schedule = "This bill is scheduled to be paid and will be paid on the due date, you're in good hands!",
  UnableToPay = 'This bill can be split into many sub-payments',
  Paid = 'Your bill has been paid',
}

export type StatusValues = keyof typeof Status;

export interface Bill {
  /**
   * unique id for the bill
   */
  id: string;
  /**
   * a thumbnail of the bill image
   */
  thumbnailUrl: string;
  /**
   * A full image for the bill
   */
  url: string;
  /**
   * Amount of the bill in AUD
   */
  amount: number;
  /**
   * Date the bill was generated in ISO format
   */
  date: Date;
  /**
   * Status of the bill
   */
  status: StatusValues;
  /**
   * Custom extra data to show on opening popup of the bill
   */
  extra: {
    statusDescription: string;
    name: string;
    paid: boolean;
  };
}

// export interface BillCollection {
//   [id: string]: Bill;
// }

/**
 * API Response error object
 */
export class ResponseError extends Error {
  public readonly status: number;

  constructor(message: string, status: number) {
    super(message);
    this.status = status;
  }

  toString() {
    return `${this.status}: ${this.message}`;
  }
}
