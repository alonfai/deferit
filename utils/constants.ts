import { Dimensions, Platform } from 'react-native';

/**
 * Different possible API Status codes
 */
export const STATUS_CODES = {
  OK: 200,
  BAD_REQUEST: 400,
  NOT_FOUND: 404,
  INTERNAL_SERVER_ERROR: 500,
};

/**
 * Get device dimensions
 */
export const { width, height } = Dimensions.get('window');

/**
 * Get device information
 */
export const Device = {
  isMobile: Platform.OS === 'android' || Platform.OS === 'ios',
  isWeb: Platform.OS === 'web',
};

/**
 * number of data entities to fetch on each API call to /photos
 */
export const LimitRequestSize = 10;

/**
 * API Fetching maximum retry attempts on error
 */
export const MAX_API_RETRIES = 3;
