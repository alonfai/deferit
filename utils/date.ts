import { getHours, getMinutes, getSeconds, getUnixTime } from 'date-fns';

export function getTimeValue(timestamp: string): string {
  const ts = new Date(timestamp);
  return `${getHours(ts).toLocaleString(undefined, { minimumIntegerDigits: 2 })}:${getMinutes(
    ts
  ).toLocaleString(undefined, { minimumIntegerDigits: 2 })}:${getSeconds(ts).toLocaleString(
    undefined,
    { minimumIntegerDigits: 2 }
  )}`;
}
