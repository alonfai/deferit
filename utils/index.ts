export * as date from './date';
export * as constants from './constants';
export * as Types from './types';
export * as helpers from './helpers';
