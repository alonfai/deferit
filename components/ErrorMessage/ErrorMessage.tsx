import React from 'react';
import { Dimensions, StyleSheet, Text, View } from 'react-native';
import { MaterialIcons } from '@expo/vector-icons';
import { Types } from '../../utils';

export type Props = {
  /**
   * Error message response from the API when unable to retrieve data
   */
  error: Types.ResponseError | null;
};

const ErrorMessage: React.FC<Props> = ({ error }) => {
  return (
    error && (
      <View style={styles.error}>
        <MaterialIcons name='error' size={24} color='red' />
        <Text style={styles.errorText}>{error.message}</Text>
      </View>
    )
  );
};
const styles = StyleSheet.create({
  error: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  errorText: {
    fontSize: 20,
    marginLeft: 10,
    maxWidth: `${Dimensions.get('window').width - 80}px`,
    maxHeight: `${Dimensions.get('window').height - 80}px`,
  },
});

export default ErrorMessage;
