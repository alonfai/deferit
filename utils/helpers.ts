import { getHours, getMinutes, getSeconds, format } from 'date-fns';
import * as faker from 'faker';

import { Types } from '../utils';

export function getTimeValue(timestamp: string): string {
  const ts = new Date(timestamp);
  return `${getHours(ts).toLocaleString(undefined, { minimumIntegerDigits: 2 })}:${getMinutes(
    ts
  ).toLocaleString(undefined, { minimumIntegerDigits: 2 })}:${getSeconds(ts).toLocaleString(
    undefined,
    { minimumIntegerDigits: 2 }
  )}`;
}

export function formatDate(date: Date, resultFormat = 'dd/MM/yyyy'): string {
  return format(date, resultFormat);
}

export function getRandomStatus(): Types.StatusValues {
  const val = faker.datatype.number({ min: 0, max: 3 });
  switch (val) {
    case 0:
      return 'Schedule';
    case 1:
      return 'Processing';
    case 2:
      return 'Paid';
    case 3:
    default:
      return 'UnableToPay';
  }
}
