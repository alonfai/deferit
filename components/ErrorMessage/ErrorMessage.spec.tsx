/// <reference types="@testing-library/jest-native" />
import React from 'react';
import { render } from '@testing-library/react-native';
import { Types } from '../../utils';

import ErrorMessage from './ErrorMessage';

describe('<ErrorFallback />', () => {
  test('Render successfully when an error object is provided', async () => {
    const error: Types.ResponseError = {
      message: 'custom message',
      name: 'error',
      status: 400,
      stack: '',
    };

    const { getByText } = render(<ErrorMessage error={error} />);
    const genericError = getByText('custom message');
    expect(genericError).toBeTruthy();
  });

  test('Render null when an error object is undefined/null', async () => {
    const { getByText } = render(<ErrorMessage error={null} />);
    const genericError = getByText('custom message');
    expect(genericError).not.toBeTruthy();
  });
});
