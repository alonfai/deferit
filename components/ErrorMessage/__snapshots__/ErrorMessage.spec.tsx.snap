// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorMessage /> Render null when an error object is undefined/null 1`] = `null`;

exports[`<ErrorMessage /> Render successfully when an error object is provided 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <Text />
  <Text
    style={
      Object {
        "fontSize": 20,
        "marginLeft": 10,
        "maxHeight": "1254px",
        "maxWidth": "670px",
      }
    }
  >
    custom message
  </Text>
</View>
`;
